font/src/ft/mod.rs

```diff
@@ -265,13 +289,18 @@ impl FreeTypeRasterizer {
                 Some(FixedSize { pixelsize: pixelsize.next().expect("has 1+ pixelsize") })
             };
 
+            // Construct harfbuzz font
+            let hb_font = Font::new(HbFace::from_file(&path, index as u32)?);
```

Comment:
You could create from ```ft_face``` like so:

```rust
    use harfbuzz_rs::HarfbuzzObject;
    use harfbuzz_rs::hb as harfbuzz_sys;
    let hb_font_raw = unsafe{harfbuzz_sys::hb_ft_font_create_referenced(ft_face)};
    let hb_font = unsafe {harfbuzz_rs::Font::from_raw(hb_font_raw)};
```
(same amount of ```unsafe``` as before).

Reply:
I agree with the sentiment, it's wasteful to create the HbFont when we already have a Ft Face ripe for the picking.

There's a frustrating issue going on here with two different freetype crates each exporting their own version of FT_Face. So if you try to convert ft_face to the expected *mut FT_FaceRec_ from freetype-0.4.1 it's actually a *mut FT_FaceRec from freetype-rs (transitively freetype-sys).  I took a look at manually converting but you end up having to recursively convert the pointers FT_Face holds and it got too gnarly.

In the long term we should upstream a variant of the harfbuzz_sys bindings that use freetype-rs types. Until then I think we're stuck with the current approach. Unless you know of a workaround I'm unaware of, that would be sweet.

Reply:
Also, I did not realize ```harfbuzz_rs``` types have a ```PhantomData``` marker so the code above wouldn't have cut it because ```Font::hb_font``` is marked ```'static```. It turns out though, current ```harfbuzz_rs``` leaks `hbface` when calling `hb_rs::Font::new`. So if we didn't want to leak  data, we'd need to the reduce reference count of the inner `hbbuff` and `hbface` of `hb_rs::Font` by implementing `Drop` on `Font`. Or avoid create `hbface`: the code below worked for me and does not leak from HB. But this quite a messy situation.

```rust
use harfbuzz_rs::HarfbuzzObject;
use harfbuzz_rs::hb as harfbuzz_sys;
let hb_font_raw = unsafe{harfbuzz_sys::hb_ft_font_create_referenced(ft_face.raw_mut() as *mut _ as *mut u64 as *mut _)};
let hb_font = unsafe {harfbuzz_rs::Font::from_raw(hb_font_raw)};

let face = Face {
    hb_font: unsafe{std::mem::transmute(hb_font)},
```

Reply:
Alright silly question, how did you get harfbuzz_sys to build with ft linked? So far I've been unsuccessful in doing so and haven't been able to figure out if it's building from source and just not linking it in or if it's my system copy and I need to recompile that.

Also, thank you for pointing out the memory leak and providing a workaround for the library incompatibility! While the pointer conversion is quite messy, it's preferable to introducing a memory leak by far.

Reply:
Ditto - That's why I did it that way in the original code

Reply:
> Alright silly question, how did you get harfbuzz_sys to build with ft linked? 

with hb-sys 0.3.2: try something like ```HARFBUZZ_SYS_NO_PKG_CONFIG=1 cargo build```. It seems possible to build on windows although I have not tested that. Also ```alpine```, ```arch``` & ```fedora 29``` seems to build hb with freetype already. If your native harbuzz comes with ```hb-ft.h``` (usually in ```/usr/include/harfbuzz```)  then it has been built with freetype support.